pyne/utils.py

```diff
@@ -190,3 +192,166 @@ def remove(path):
+    of float decimals.
+    Return False if two strings are different.
+    """
+    # if string can be converted to float number
+    if is_float(s1) and is_float(s2):
+        return is_close(float(s1), float(s2), rel_tol)
+    else:
+        # not a number
+        return s1 == s2
+
+
+def line_almost_same(l1, l2, rel_tol=1e-9):
+    """
+    This function is used to compare two lines (read from files). If they are
+    the same, or almost the same (with only silight different on float
+    numbers), return True. Ohterwise, return False.
+
+    Parameters:
+    -----------
+    l1 : str
+        Line 1
+    l2 : str
+        Line 2
+    rel_tol : float
+        Reletive tolerance for float comparison
+
+    Returns:
+    --------
+    True, if two lines are the same. False, if they are different.
+    """
+    if l1 == l2:
+        # exactly the same
+        return True
+    else:
+        # There are differences
+        ls1 = l1.strip().split()
+        ls2 = l2.strip().split()
+        if len(ls1) != len(ls2):
+            return False
+        else:
+            # compare string elements of the line
+            for i in range(len(ls1)):
+                if str_almost_same(ls1[i], ls2[i], rel_tol):
+                    pass
+                else:
+                    return False
+        return True
+
+
+def file_almost_same(f1, f2, rel_tol=1e-9):
+    """
+    For some reasones, it's useful to compare two files that are almost the
+    same. Two files, f1 and f2, the text contents are exactly the same, but
+    there is a small difference in numbers. Such as the difference between
+    'some text 9.5' and 'some text 9.500000000001'.
+    For example, in PyNE test files, there are some expected file generated
+    by python2, however, the the file generated by python3 may have difference
+    in dicimals.
+
+    Parameters:
+    -----------
+    f1 : str 
+        Filename of file 1 or lines
+    f2 : str
+        Filename of file 2 or lines
+    rel_tol : float
+        Reletive tolerance for float numbers
+
+    Returns:
+    True : bool
+        If two file are exactly the same, or almost the same with only dicimal
+        differences.
+    False : bool
+        If the difference the two files are more than dicimal differences.
+    """
+    if os.path.isfile(f1) and os.path.isfile(f2):
+        if filecmp.cmp(f1, f2):
+            # precheck
+            return True
+    else:
+        # read lines of f1 and f2, convert to unicode
+        if os.path.isfile(f1):
+            with open(f1, 'r') as f:
+                lines1 = f.readlines()
+                lines1 = str_to_unicode(lines1)
+        else:
+            lines1 = str_to_unicode(f1)
+        lines1 = lines1.strip().split(u'\n')
+
+        if os.path.isfile(f2):
+            with open(f2, 'r') as f:
+                lines2 = f.readlines()
+                lines2 = str_to_unicode(lines2)
+        else:
+            lines2 = str_to_unicode(f2)
```

Comment:
maybe you could build a method to return file's lines converted to unicode to avoid redundancy ?

Reply:
I ma not 100% convince about this comment, I don't like the redundancy, but it only occurs once and this is very specific...

Reply:
Put the `str_to_unicode()` outside the `if/else` block would reduce the redundancy.